pipenv/project.py

```diff
@@ -628,15 +633,35 @@ def sources(self):
         else:
             return [DEFAULT_SOURCE]
 
+    def find_source(self, source):
+        """given a source, find it.
+
+        source can be a url or an index name.
+        """
+        if not is_valid_url(source):
+            try:
+                source = self.get_source(name=source)
+            except SourceNotFound:
+                source = self.get_source(url=source)
+        else:
+            source = self.get_source(url=source)
+        return source
+
     def get_source(self, name=None, url=None):
-        for source in self.sources:
+        def find_source(sources, name=None, url=None):
             if name:
-                if source.get('name') == name:
-                    return source
-
+                source = [s for s in sources if s.get('name') == name]
             elif url:
-                if source.get('url') in url:
-                    return source
+                source = [s for s in sources if s.get('url') in url]
+            if source:
+                return source[0]
```

Comment:
Isn’t this essentially what `first` does? I don’t dislike this (I never liked the `first` library, in fact), but maybe we should be consistent here.

Reply:
if it's imported I'll swap it over, in the spirit of speeding up startup time (which primarily was coming from indirect loading of the project module) I was trying to avoid extra imports